daemon/commands.c

```diff
@@ -0,0 +1,539 @@
+}
+
+static cmd_status_t cmd_exit_execute(char *args, char **message)
+{
+    (void)args;
+    (void)message;
+
+    error_log_limit_unlimited();
+    info("COMMAND: Cleaning up to exit.");
+    netdata_cleanup_and_exit(0);
+    exit(0);
+
+    return CMD_STATUS_SUCCESS;
+}
+
+static cmd_status_t cmd_fatal_execute(char *args, char **message)
+{
+    (void)args;
+    (void)message;
+
+    fatal("COMMAND: netdata now exits.");
+
+    return CMD_STATUS_SUCCESS;
+}
+
+static void cmd_lock_exclusive(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_wrlock(&exclusive_rwlock);
+}
+
+static void cmd_lock_orthogonal(unsigned index)
+{
+    uv_rwlock_rdlock(&exclusive_rwlock);
+    uv_mutex_lock(&command_lock_array[index]);
+}
+
+static void cmd_lock_idempotent(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_rdlock(&exclusive_rwlock);
+}
+
+static void cmd_lock_high_priority(unsigned index)
+{
+    (void)index;
+}
+
+static void cmd_unlock_exclusive(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_wrunlock(&exclusive_rwlock);
+}
+
+static void cmd_unlock_orthogonal(unsigned index)
+{
+    uv_rwlock_rdunlock(&exclusive_rwlock);
+    uv_mutex_unlock(&command_lock_array[index]);
+}
+
+static void cmd_unlock_idempotent(unsigned index)
+{
+    (void)index;
+
+    uv_rwlock_rdunlock(&exclusive_rwlock);
+}
+
+static void cmd_unlock_high_priority(unsigned index)
+{
+    (void)index;
+}
+
+static void pipe_write_cb(uv_write_t* req, int status)
+{
+    (void)status;
+    uv_pipe_t *client = req->data;
+
+    uv_close((uv_handle_t *)client, NULL);
+    freez(client);
+    --clients;
+    info("Command Clients = %u\n", clients);
+}
+
+static inline void add_string_to_command_reply(char *reply_string, unsigned *reply_string_size, char *str)
+{
+    unsigned len;
+
+    len = strlen(str);
+    strncpyz(reply_string + *reply_string_size, str, len);
+    *reply_string_size += len;
+}
+
+static void send_command_reply(uv_stream_t *client, cmd_status_t status, char *message)
+{
+    int ret;
+    char reply_string[MAX_COMMAND_LENGTH] = {'\0', };
+    unsigned reply_string_size = 0;
+    uv_buf_t write_buf;
+    uv_write_t write_req;
+
+    add_string_to_command_reply(reply_string, &reply_string_size, cmd_status_str[status]);
+    if (message) {
+        add_string_to_command_reply(reply_string, &reply_string_size, ": ");
+        add_string_to_command_reply(reply_string, &reply_string_size, message);
+    }
+
+    write_req.data = client;
+    write_buf.base = reply_string;
+    write_buf.len = strlen((char *)write_buf.base);
+    ret = uv_write(&write_req, (uv_stream_t *)client, &write_buf, 1, pipe_write_cb);
+    if (ret) {
+        error("uv_write(): %s", uv_strerror(ret));
+    }
+    info("COMMAND: Sending reply: \"%s\"", reply_string);
+}
+
+cmd_status_t execute_command(cmd_t idx, char *args, char **message)
+{
+    cmd_status_t status;
+    cmd_type_t type = command_info_array[idx].type;
+
+    cmd_lock_by_type[type](idx);
```

Comment:
Having mutexes on each command makes sense, as it stops a collision with the signal handling code in the daemon. This doesn't make any sense to me. Why do we want to stop two idempotent commands from running at the same time, or two high-priority commands.. ?
The one case it would make sense for if the command type is exclusive, it does not make sense for the the types.

Reply:
It's a reader-writer lock. The commands are not stopped because they are readers. The only writer is a command of exclusive type . The mutexes exist only for orthogonal commands and they are command specific.